numpy/lib/shape_base.py

```diff
@@ -16,10 +16,236 @@
 __all__ = [
     'column_stack', 'row_stack', 'dstack', 'array_split', 'split',
     'hsplit', 'vsplit', 'dsplit', 'apply_over_axes', 'expand_dims',
-    'apply_along_axis', 'kron', 'tile', 'get_array_wrap'
+    'apply_along_axis', 'kron', 'tile', 'get_array_wrap', 'take_along_axis',
+    'put_along_axis'
     ]
 
 
+def _make_along_axis_idx(arr_shape, indices, axis):
```

Comment:
I had considered asking you to just pass on the shape, but thought it was too much trouble, but now I see it ended up being needed! Nicer.

Reply:
I also considered adding `.shape ` and `.ndim` to flatiter, but then I'd need to do it to `MaskedArray.flat` too, and it was a little unclear whether .shape and .ndim should refer to the flattened or original array.